madara/crates/client/settlement_client/src/eth/event.rs

```diff
@@ -0,0 +1,286 @@
+use crate::error::SettlementClientError;
+use crate::eth::error::EthereumClientError;
+use crate::eth::StarknetCoreContract::LogMessageToL2;
+use crate::messaging::L1toL2MessagingEventData;
+use crate::utils::u256_to_felt;
+use alloy::contract::EventPoller;
+use alloy::rpc::types::Log;
+use alloy::transports::http::{Client, Http};
+use futures::Stream;
+use starknet_types_core::felt::Felt;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+type EthereumStreamItem = Result<(LogMessageToL2, Log), alloy::sol_types::Error>;
+type EthereumStreamType = Pin<Box<dyn Stream<Item = EthereumStreamItem> + Send + 'static>>;
+
+pub struct EthereumEventStream {
+    pub stream: EthereumStreamType,
+}
+
+impl EthereumEventStream {
+    pub fn new(watcher: EventPoller<Http<Client>, LogMessageToL2>) -> Self {
+        let stream = watcher.into_stream();
+        Self { stream: Box::pin(stream) }
+    }
+}
+
+impl Stream for EthereumEventStream {
+    type Item = Result<L1toL2MessagingEventData, SettlementClientError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.stream.as_mut().poll_next(cx) {
+            Poll::Ready(Some(result)) => match result {
+                Ok((event, log)) => {
+                    let event_data = (|| -> Result<L1toL2MessagingEventData, SettlementClientError> {
+                        Ok(L1toL2MessagingEventData {
```

Comment:
Consider implementing `TryFrom` for this, it would make `poll_next()` easier to grok